### ISSUE TITLE
Add Offical Rust Bindings

### ISSUE THREAD

**Issue by Snowiiii** [KhronosGroup/Vulkan-Docs#2268]
The currently best solution for Vulkan bindings is [Ash](https://github.com/ash-rs/ash). Ash is a fine library, but it does have some suboptimal design choices, in my opinion.

First, it's not entirely accurate to say that everything is generated and it's just like working in C or something. Ash introduces its own concepts, such as a custom [device](https://github.com/ash-rs/ash/blob/master/ash/src/device.rs). In this device module, you can find many functions derived from Ash, but not all of them are complete, and some crucial features are missing. The project structure could use some improvement as well. There have been efforts to rewrite the Ash core and the generator for the purpose of addressing design issues, but progress on these files has been stagnant for several months.

The ultimate goal is to have official Rust bindings developed under the Khronos Group's guidance. This would not only attract more developers to work on these bindings but also encourage more developers to use them.

It's important to note that most members of the Khronos Group may not have experience working with Rust, and they might be content with sticking to C/C++. However, it's worth highlighting that providing bindings for Rust is often much easier compared to languages like Java or Python.

In conclusion, I don't want to convey that Ash is a bad solution. It's a valuable library. I simply believe that official support could bring many advantages, including the opportunity to enhance the library's design from the beginning, rather than relying on lengthy rewrites.

**Comment by MarijnS95** [comment]
> First, it's not entirely accurate to say that everything is generated and it's just like working in C or something. Ash introduces its own concepts, such as a custom [device](https://github.com/ash-rs/ash/blob/master/ash/src/device.rs). In this device module, you can find many functions derived from Ash, but not all of them are complete, and some crucial features are missing. The project structure could use some improvement as well. There have been efforts to rewrite the Ash core and the generator for the purpose of addressing design issues, but progress on these files has been stagnant for several months.

Perhaps it might be worth bringing this (and especially those gripes in more detail!) up with the Ash maintainers - or better yet, help them make progress on these issues - rather than enticing Khronos to build separate new Rust bindings without their knowledge. Disclaimer: I'm one of those maintainers.

Conveniently we're already working with Khronos by e.g. having our code generator be part of the CI system here and in the upstream repo to make sure `vk.xml` can consistently be used to generate Rust code. And we're listed as the language bindings for Rust: https://www.vulkan.org/tools#language-bindings (the link still points to an older repo though, at least GitHub redirects it).

If there is enough community/commercial interest a working group sounds like a good idea, while making sure that industry expertise/experts in both Rust and Vulkan are involved (i.e. the authors and maintainers of the popular bindings/graphics crates, especially since it was said Khronos does not currently have the necessary expertise for Rust). A working group requires carefully analyzing and documenting what is currently there, and what is missing or broken, before even being able to justify creating something new. It is the right place to discuss biased design choices (instead of calling them suboptimal design choices :wink:), and I believe that Ash has enough traction in the community to stick with improving that rather than forcing everyone onto a new "official" bindings crate.
(For example, we're all seeing how slow the transition from `winapi` to Microsoft's _official_ `windows-rs` bindings is... And `winapi` hasn't been maintained for over 3 years!)

Rewrites don't have to be lengthy; they just **don't make progress** because the vast majority is happy with the current look and feel of Ash! And the planned "generator rewrite" is only to aid maintainers in generating slightly-higher-level Version/Extension wrappers, instead of handwriting them. And that will pose to be **super** tricky as there's a lot of thinking and manual analyzing of various documentation sources to write them, that isn't simply conveyed in `vk.xml`.

---

This issue isn't the right place to answer questions about Ash, but I do want to set some expectations and views straight for future readers:

> First, it's not entirely accurate to say that everything is generated 

**All** of Vulkan has an autogenerated binding in Ash and can be used. Anything that's missing or written by hand are higher-level helpers on top that one _is not required_ to use. Not everything _can_ be generated, as expressed above certain things require manual intervention.
(I stumbled upon this issue when wanting to file a new one about `vkGetLatencyTimingsNV` + `VkGetLatencyMarkerInfoNV` being super inconvenient to use from a Rust perspective, _even if_ handwriting the bindings)

> and it's just like working in C or something

That is by design/intent. Ash aims to map the C bindings to Rust with very minimal Rust'y interfaces sprinkled on top (slices, borrows, use of `Option` and `Result`). If you are looking for something more high-level that aims to wrap Vulkan in a lifetime- and object-safe way, consider: https://crates.io/crates/vulkano (built on top of the _low-level_ bindings provided by Ash).

> Ash introduces its own concepts, such as a custom [device](https://github.com/ash-rs/ash/blob/master/ash/src/device.rs).

This - together with the hand-written extension wrappers - is one of the areas where Ash protrudes its `-ffi`/`-sys`'ness. There have been discussions about splitting the `ash` crate in two to "solve" this for those that desire exclusively pure `ffi` bindings (effectively nothing more than type definitions), but a verdict never settled.

However, the request for Rust Vulkan bindings to be less "like working in C" contradict the request to not "introduce its own concepts" that harness the power and flexibility of Rust in a better way?

> you can find many functions derived from Ash, but not all of them are complete, and some crucial features are missing.

I would be very curious to hear exactly what is missing from our Vulkan 1.0, 1.1, 1.2 and 1.3 wrapper functions. We've done a pass some time ago to make sure all are provided. Backing up this statement is a good way to drive the point home.

**Comment by oddhack** [comment]
@Snowiiii I know this isn't the answer you wanted, but as @TomOlson noted, this isn't something the Vulkan WG has the expertise or focus to take on. Please close this unless there's something else actionable for us to do.
